src/classification/utils.py

```diff
@@ -0,0 +1,147 @@
+#!/usr/bin/env python
+# coding: utf-8
+
+""" Utility functions
+"""
+
+import random
+import tarfile
+import typing as tp
+
+import braceexpand
+import numpy as np
+import timm
+import torch
+from timm.scheduler import CosineLRScheduler
+
+from src.classification.constants import (
+    AVAILABLE_MODELS,
+    COSINE_LR_SCHEDULER,
+    CROSS_ENTROPY_LOSS,
+    VIT_B16_128,
+)
+
+
+def set_random_seeds(random_seed: int) -> None:
+    """Set random seeds for reproducibility"""
+
+    random.seed(random_seed)
+    np.random.seed(random_seed)
+    torch.manual_seed(random_seed)
+    torch.cuda.manual_seed(random_seed)
+    torch.backends.cudnn.deterministic = True
+
+
+def get_optimizer(
+    optimizer_type: str,
+    model: torch.nn.Module,
+    learning_rate: float,
+    weight_decay: float,
+    momentum: float = 0.9,
+) -> torch.optim.Optimizer:
+    """Optimizer definitions"""
+
+    if optimizer_type == "adamw":
+        return torch.optim.AdamW(
+            model.parameters(), lr=learning_rate, weight_decay=weight_decay
+        )
+    elif optimizer_type == "sgd":
+        return torch.optim.SGD(
+            model.parameters(),
+            lr=learning_rate,
+            momentum=momentum,
+            weight_decay=weight_decay,
+        )
+    else:
+        raise RuntimeError(f"{optimizer_type} optimizer is not implemented.")
+
+
+def get_learning_rate_scheduler(
+    optimizer: torch.optim.Optimizer,
+    lr_scheduler_type: str,
+    total_epochs: int,
+    steps_per_epoch: int,
+    warmup_epochs: int,
+) -> tp.Any:
+    """Learning rate scheduler definitions"""
+
+    total_steps = int(total_epochs * steps_per_epoch)
+    warmup_steps = int(warmup_epochs * steps_per_epoch)
+
+    if lr_scheduler_type == COSINE_LR_SCHEDULER:
+        return CosineLRScheduler(
+            optimizer,
+            t_initial=(total_steps - warmup_steps),
+            warmup_t=warmup_steps,
+            warmup_prefix=True,
+            cycle_limit=1,
+            t_in_epochs=False,
+        )
+    else:
+        raise RuntimeError(
+            f"{lr_scheduler_type} learning rate scheduler is not implemented."
+        )
+
+
+def get_loss_function(
+    loss_function_name: str, label_smoothing: float = 0.0
+) -> torch.nn.Module:
+    """Loss function definitions"""
+
+    if loss_function_name == CROSS_ENTROPY_LOSS:
+        return torch.nn.CrossEntropyLoss(label_smoothing=label_smoothing)
+    else:
+        raise RuntimeError(f"{loss_function_name} loss is not implemented.")
```

Comment:
Ideally, these elses should also be treated like the other examples.

For example:
```python
if lr_scheduler_type != COSINE_LR_SCHEDULER:
    raise RuntimeError()

scheduler = CosineLRScheduler(
                optimizer,
                t_initial=(total_steps - warmup_steps),
                warmup_t=warmup_steps,
                warmup_prefix=True,
                cycle_limit=1,
                t_in_epochs=False,
    )
return scheduler
```

That way, your code is already ready for more cases (other scheduler types or loss functions), with no extra cost or complexity in the code

Reply:
Thanks, Francis, for this comment! I'll update it in another PR, otherwise, I won't be able to merge this one.